File: code-tiny/tools/servlet_jsp/el_parser.py
from __future__ import annotations

from dataclasses import dataclass
from typing import List, Optional, Tuple

from tools.servlet_jsp.models import Diagnostic, ResourceBudgets, SourceSpan


_KEYWORDS = {
    "and",
    "div",
    "empty",
    "eq",
    "false",
    "ge",
    "gt",
    "instanceof",
    "le",
    "lt",
    "mod",
    "ne",
    "not",
    "null",
    "or",
    "true",
}

_IMPLICIT_SCOPES = {
    "param": "parameter",
    "paramValues": "parameter",
    "requestScope": "request",
    "sessionScope": "session",
    "applicationScope": "application",
    "cookie": "cookie",
    "header": "header",
    "headerValues": "header",
}


@dataclass(frozen=True)
class ELToken:
    kind: str
    value: str
    start_offset: int
    end_offset: int


@dataclass(frozen=True)
class ELFunctionReference:
    prefix: str
    name: str
    raw: str
    start_offset: int
    end_offset: int


@dataclass(frozen=True)
class ELReference:
    root: str
    path: Tuple[str, ...]
    raw: str
    start_offset: int
    end_offset: int
    resolution_status: str = "unresolved"
    implicit_scope: str = ""

    @property
    def property_path(self) -> str:
        value = self.root
        for part in self.path:
            value += part if part.startswith("[") else f".{part}"
        return value


@dataclass(frozen=True)
class ELStateRead:
    implicit_object: str
    scope: str
    name: str
    raw: str
    dynamic: bool
    start_offset: int
    end_offset: int


@dataclass(frozen=True)
class ELParseResult:
    raw: str
    body: str
    span: SourceSpan
    tokens: Tuple[ELToken, ...] = ()
    references: Tuple[ELReference, ...] = ()
    functions: Tuple[ELFunctionReference, ...] = ()
    state_reads: Tuple[ELStateRead, ...] = ()
    diagnostics: Tuple[Diagnostic, ...] = ()
    truncated: bool = False
    complete: bool = True

    @property
    def variables(self) -> Tuple[str, ...]:
        return tuple(dict.fromkeys(item.root for item in self.references))

    @property
    def property_paths(self) -> Tuple[str, ...]:
        return tuple(dict.fromkeys(item.property_path for item in self.references))

    @property
    def implicit_objects(self) -> Tuple[str, ...]:
        return tuple(dict.fromkeys(item.implicit_object for item in self.state_reads))


def parse_el_expression(
    expression: str,
    *,
    file_path: str = "",
    start_line: int = 1,
    start_column: int = 1,
    budgets: Optional[ResourceBudgets] = None,
) -> ELParseResult:
    """Extract static references from one JSP EL expression without evaluating it."""

    effective = budgets or ResourceBudgets()
    raw = expression or ""
    diagnostics: List[Diagnostic] = []
    truncated = False
    encoded = raw.encode("utf-8", errors="surrogatepass")
    if len(encoded) > effective.max_el_bytes:
        bounded = encoded[: effective.max_el_bytes]
        while bounded:
            try:
                scanned = bounded.decode("utf-8", errors="strict")
                break
            except UnicodeDecodeError as exc:
                bounded = bounded[: exc.start]
        else:
            scanned = ""
        truncated = True
        diagnostics.append(
            Diagnostic(
                "servlet_jsp.el.byte_budget",
                f"EL byte budget {effective.max_el_bytes} reached",
                "warning",
                file_path,
                start_line,
                start_line,
            )
        )
    else:
        scanned = raw

    body_offset = 0
    complete = True
    if scanned.startswith(("${", "#{")):
        body_offset = 2
        if scanned.endswith("}"):
            body = scanned[2:-1]
        else:
            body = scanned[2:]
            complete = False
            diagnostics.append(
                Diagnostic(
                    "servlet_jsp.el.unclosed",
                    "Unclosed EL expression",
                    "warning",
                    file_path,
                    start_line,
                    start_line,
                )
            )
    else:
        body = scanned

    tokens, token_diagnostics, token_truncated, token_complete = _lex(
        body,
        body_offset=body_offset,
        file_path=file_path,
        start_line=start_line,
        max_tokens=effective.max_el_tokens,
        max_nesting=effective.max_el_nesting,
    )
    diagnostics.extend(token_diagnostics)
    truncated = truncated or token_truncated
    functions, function_token_indexes = _function_references(tokens, scanned)
    references = _references(tokens, scanned, function_token_indexes)
    state_reads = _state_reads(references)
    span = SourceSpan(
        file_path=file_path,
        start_line=start_line,
        end_line=start_line + raw.count("\n"),
        start_column=start_column,
        end_column=start_column + len(raw) if "\n" not in raw else 1,
    )
    return ELParseResult(
        raw=raw,
        body=body,
        span=span,
        tokens=tuple(tokens),
        references=tuple(references),
        functions=tuple(functions),
        state_reads=tuple(state_reads),
        diagnostics=tuple(diagnostics),
        truncated=truncated,
        complete=complete and token_complete and not truncated,
    )


def _lex(
    body: str,
    *,
    body_offset: int,
    file_path: str,
    start_line: int,
    max_tokens: int,
    max_nesting: int,
) -> Tuple[List[ELToken], List[Diagnostic], bool, bool]:
    tokens: List[ELToken] = []
    diagnostics: List[Diagnostic] = []
    index = 0
    nesting: List[str] = []
    truncated = False
    complete = True
    while index < len(body):
        char = body[index]
        if char.isspace():
            index += 1
            continue
        if body.startswith("//", index):
            end = body.find("\n", index + 2)
            index = len(body) if end < 0 else end + 1
            continue
        if body.startswith("/*", index):
            end = body.find("*/", index + 2)
            if end < 0:
                complete = False
                diagnostics.append(
                    Diagnostic(
                        "servlet_jsp.el.unclosed_comment",
                        "Unclosed comment in EL expression",
                        "warning",
                        file_path,
                        start_line,
                        start_line,
                    )
                )
                break
            index = end + 2
            continue
        start = index
        if char in {"'", '"'}:
            quote = char
            index += 1
            escaped = False
            while index < len(body):
                current = body[index]
                index += 1
                if escaped:
                    escaped = False
                elif current == "\\":
                    escaped = True
                elif current == quote:
                    break
            else:
                complete = False
                diagnostics.append(
                    Diagnostic(
                        "servlet_jsp.el.unclosed_string",
                        "Unclosed string literal in EL expression",
                        "warning",
                        file_path,
                        start_line,
                        start_line,
                    )
                )
            kind = "string"
        elif char.isalpha() or char in {"_", "$"}:
            index += 1
            while index < len(body) and (body[index].isalnum() or body[index] in {"_", "$"}):
                index += 1
            kind = "identifier"
        elif char.isdigit():
            index += 1
            while index < len(body) and (body[index].isalnum() or body[index] in {".", "_"}):
                index += 1
            kind = "number"
        else:
            pair = body[index : index + 2]
            if pair in {"==", "!=", "<=", ">=", "&&", "||", "->", "+=", "-=", "*=", "/=", "?."}:
                index += 2
            else:
                index += 1
            kind = "punctuation" if char in "()[]{}.,:?" else "operator"
        value = body[start:index]
        tokens.append(ELToken(kind, value, start + body_offset, index + body_offset))
        if value in {"(", "[", "{"}:
            nesting.append(value)
            if len(nesting) > max_nesting:
                diagnostics.append(
                    Diagnostic(
                        "servlet_jsp.el.nesting_budget",
                        f"EL nesting budget {max_nesting} reached",
                        "warning",
                        file_path,
                        start_line,
                        start_line,
                    )
                )
                truncated = True
                break
        elif value in {")", "]", "}"}:
            expected = {")": "(", "]": "[", "}": "{"}[value]
            if not nesting or nesting[-1] != expected:
                complete = False
                diagnostics.append(
                    Diagnostic(
                        "servlet_jsp.el.unexpected_closer",
                        f"Unexpected closing token {value!r} in EL expression",
                        "warning",
                        file_path,
                        start_line,
                        start_line,
                    )
                )
            else:
                nesting.pop()
        if len(tokens) >= max_tokens:
            if index < len(body):
                diagnostics.append(
                    Diagnostic(
                        "servlet_jsp.el.token_budget",
                        f"EL token budget {max_tokens} reached",
                        "warning",
                        file_path,
                        start_line,
                        start_line,
                    )
                )
                truncated = True
            break
    if nesting and not truncated:
        complete = False
        diagnostics.append(
            Diagnostic(
                "servlet_jsp.el.unbalanced_nesting",
                "Unbalanced grouping token in EL expression",
                "warning",
                file_path,
                start_line,
                start_line,
            )
        )
    return tokens, diagnostics, truncated, complete


def _function_references(
    tokens: List[ELToken], raw: str
) -> Tuple[List[ELFunctionReference], set[int]]:
    functions: List[ELFunctionReference] = []
    indexes: set[int] = set()
    for index in range(len(tokens) - 3):
        prefix, colon, name, opening = tokens[index : index + 4]
        if (
            prefix.kind == "identifier"
            and colon.value == ":"
            and name.kind == "identifier"
            and opening.value == "("
        ):
            indexes.update({index, index + 2})
            functions.append(
                ELFunctionReference(
                    prefix=prefix.value,
                    name=name.value,
                    raw=raw[prefix.start_offset : name.end_offset],
                    start_offset=prefix.start_offset,
                    end_offset=name.end_offset,
                )
            )
    return functions, indexes


def _references(
    tokens: List[ELToken], raw: str, function_token_indexes: set[int]
) -> List[ELReference]:
    references: List[ELReference] = []
    for index, token in enumerate(tokens):
        if token.kind != "identifier" or token.value in _KEYWORDS or index in function_token_indexes:
            continue
        previous = tokens[index - 1].value if index else ""
        if previous in {".", "?.", ":"}:
            continue
        path: List[str] = []
        end_offset = token.end_offset
        cursor = index + 1
        dynamic = False
        while cursor < len(tokens):
            marker = tokens[cursor]
            if marker.value in {".", "?."} and cursor + 1 < len(tokens) and tokens[cursor + 1].kind == "identifier":
                part = tokens[cursor + 1]
                path.append(part.value)
                end_offset = part.end_offset
                cursor += 2
                continue
            if marker.value == "[":
                close = _matching_bracket(tokens, cursor)
                if close is None:
                    dynamic = True
                    end_offset = marker.end_offset
                    break
                inner = tokens[cursor + 1 : close]
                if len(inner) == 1 and inner[0].kind in {"string", "number"}:
                    index_value = _literal_index(inner[0])
                    path.append(f"[{index_value}]")
                else:
                    raw_inner = raw[marker.end_offset : tokens[close].start_offset].strip()
                    path.append(f"[{raw_inner}]")
                    dynamic = True
                end_offset = tokens[close].end_offset
                cursor = close + 1
                continue
            break
        implicit_scope = _IMPLICIT_SCOPES.get(token.value, "")
        references.append(
            ELReference(
                root=token.value,
                path=tuple(path),
                raw=raw[token.start_offset:end_offset],
                start_offset=token.start_offset,
                end_offset=end_offset,
                resolution_status=(
                    "resolved" if implicit_scope and path and not dynamic else "dynamic" if dynamic else "unresolved"
                ),
                implicit_scope=implicit_scope,
            )
        )
    return references


def _matching_bracket(tokens: List[ELToken], opening: int) -> Optional[int]:
    depth = 0
    for index in range(opening, len(tokens)):
        if tokens[index].value == "[":
            depth += 1
        elif tokens[index].value == "]":
            depth -= 1
            if depth == 0:
                return index
    return None


def _literal_index(token: ELToken) -> str:
    if token.kind == "string" and len(token.value) >= 2:
        return token.value[1:-1]
    return token.value


def _state_reads(references: List[ELReference]) -> List[ELStateRead]:
    reads: List[ELStateRead] = []
    for reference in references:
        if not reference.implicit_scope:
            continue
        name = ""
        dynamic = True
        if reference.path:
            first = reference.path[0]
            if first.startswith("[") and first.endswith("]"):
                name = first[1:-1]
            else:
                name = first
            dynamic = not bool(name) or any(char in name for char in "${}[]()")
        reads.append(
            ELStateRead(
                implicit_object=reference.root,
                scope=reference.implicit_scope,
                name=name,
                raw=reference.raw,
                dynamic=dynamic,
                start_offset=reference.start_offset,
                end_offset=reference.end_offset,
            )
        )
    return reads

File: rust/crates/analyzer-jvm-overlays/src/servlet_jsp/el_parser.rs
//! Port `tools/servlet_jsp/el_parser.py` — EL tokenizer + references.

use std::collections::BTreeSet;

use crate::servlet_jsp::models::{Diagnostic, ResourceBudgets, SourceSpan};

const KEYWORDS: [&str; 16] = [
    "and", "div", "empty", "eq", "false", "ge", "gt", "instanceof", "le", "lt", "mod", "ne", "not", "null", "or",
    "true",
];

fn implicit_scope_of(name: &str) -> &'static str {
    match name {
        "param" | "paramValues" => "parameter",
        "requestScope" => "request",
        "sessionScope" => "session",
        "applicationScope" => "application",
        "cookie" => "cookie",
        "header" | "headerValues" => "header",
        _ => "",
    }
}

#[derive(Debug, Clone)]
pub struct ElToken {
    pub kind: String,
    pub value: String,
    pub start_offset: i64,
    pub end_offset: i64,
}

#[derive(Debug, Clone)]
pub struct ElFunctionReference {
    pub prefix: String,
    pub name: String,
    pub raw: String,
    pub start_offset: i64,
    pub end_offset: i64,
}

#[derive(Debug, Clone)]
pub struct ElReference {
    pub root: String,
    pub path: Vec<String>,
    pub raw: String,
    pub start_offset: i64,
    pub end_offset: i64,
    pub resolution_status: String,
    pub implicit_scope: String,
}

impl ElReference {
    pub fn property_path(&self) -> String {
        let mut value = self.root.clone();
        for part in &self.path {
            if part.starts_with('[') {
                value.push_str(part);
            } else {
                value.push('.');
                value.push_str(part);
            }
        }
        value
    }
}

#[derive(Debug, Clone)]
pub struct ElStateRead {
    pub implicit_object: String,
    pub scope: String,
    pub name: String,
    pub raw: String,
    pub dynamic: bool,
    pub start_offset: i64,
    pub end_offset: i64,
}

#[derive(Debug, Clone, Default)]
pub struct ElParseResult {
    pub raw: String,
    pub body: String,
    pub span: Option<SourceSpan>,
    pub tokens: Vec<ElToken>,
    pub references: Vec<ElReference>,
    pub functions: Vec<ElFunctionReference>,
    pub state_reads: Vec<ElStateRead>,
    pub diagnostics: Vec<Diagnostic>,
    pub truncated: bool,
    pub complete: bool,
}

impl ElParseResult {
    pub fn variables(&self) -> Vec<String> {
        let mut seen: BTreeSet<String> = BTreeSet::new();
        let mut out: Vec<String> = Vec::new();
        for item in &self.references {
            if seen.insert(item.root.clone()) {
                out.push(item.root.clone());
            }
        }
        out
    }

    pub fn property_paths(&self) -> Vec<String> {
        let mut seen: BTreeSet<String> = BTreeSet::new();
        let mut out: Vec<String> = Vec::new();
        for item in &self.references {
            let path = item.property_path();
            if seen.insert(path.clone()) {
                out.push(path);
            }
        }
        out
    }

    pub fn implicit_objects(&self) -> Vec<String> {
        let mut seen: BTreeSet<String> = BTreeSet::new();
        let mut out: Vec<String> = Vec::new();
        for item in &self.state_reads {
            if seen.insert(item.implicit_object.clone()) {
                out.push(item.implicit_object.clone());
            }
        }
        out
    }
}

pub fn parse_el_expression(
    expression: &str,
    file_path: &str,
    start_line: i64,
    start_column: i64,
    budgets: &ResourceBudgets,
) -> ElParseResult {
    let raw = expression.to_string();
    let mut diagnostics: Vec<Diagnostic> = Vec::new();
    let mut truncated = false;
    let scanned = raw.clone();
    if raw.len() > budgets.max_el_bytes as usize {
        truncated = true;
        diagnostics.push(Diagnostic::new(
            "servlet_jsp.el.byte_budget",
            &format!("EL byte budget {} reached", budgets.max_el_bytes),
            "warning",
            file_path,
            start_line,
            start_line,
        ));
    }

    let mut body_offset = 0i64;
    let mut complete = true;
    let body = if scanned.starts_with("${") || scanned.starts_with("#{") {
        body_offset = 2;
        if scanned.ends_with('}') {
            scanned[2..scanned.len() - 1].to_string()
        } else {
            complete = false;
            diagnostics.push(Diagnostic::new(
                "servlet_jsp.el.unclosed",
                "Unclosed EL expression",
                "warning",
                file_path,
                start_line,
                start_line,
            ));
            scanned[2..].to_string()
        }
    } else {
        scanned.clone()
    };

    let (tokens, token_diagnostics, token_truncated, token_complete) = lex(
        &body,
        body_offset,
        file_path,
        start_line,
        budgets.max_el_tokens as usize,
        budgets.max_el_nesting as usize,
    );
    diagnostics.extend(token_diagnostics);
    let truncated = truncated || token_truncated;
    let (functions, function_token_indexes) = function_references(&tokens, &scanned);
    let references = references(&tokens, &scanned, &function_token_indexes);
    let state_reads = state_reads(&references);
    let span = SourceSpan {
        file_path: file_path.to_string(),
        start_line,
        end_line: start_line + raw.matches('\n').count() as i64,
        start_column,
        end_column: if !raw.contains('\n') {
            start_column + raw.chars().count() as i64
        } else {
            1
        },
    };
    let complete = complete && token_complete && !truncated;
    ElParseResult {
        raw,
        body,
        span: Some(span),
        tokens,
        references,
        functions,
        state_reads,
        diagnostics,
        truncated,
        complete,
    }
}

fn lex(
    body: &str,
    body_offset: i64,
    file_path: &str,
    start_line: i64,
    max_tokens: usize,
    max_nesting: usize,
) -> (Vec<ElToken>, Vec<Diagnostic>, bool, bool) {
    let mut tokens: Vec<ElToken> = Vec::new();
    let mut diagnostics: Vec<Diagnostic> = Vec::new();
    let mut truncated = false;
    let mut complete = true;
    let chars: Vec<char> = body.chars().collect();
    let mut index = 0usize;
    let mut nesting: Vec<char> = Vec::new();
    while index < chars.len() {
        let ch = chars[index];
        if ch.is_whitespace() {
            index += 1;
            continue;
        }
        let rest: String = chars[index..(index + 2).min(chars.len())].iter().collect();
        if rest == "//" {
            // line comment
            let mut cursor = index + 2;
            while cursor < chars.len() && chars[cursor] != '\n' {
                cursor += 1;
            }
            index = if cursor >= chars.len() { chars.len() } else { cursor + 1 };
            continue;
        }
        if rest == "/*" {
            let mut cursor = index + 2;
            let mut found = false;
            while cursor + 1 < chars.len() {
                if chars[cursor] == '*' && chars[cursor + 1] == '/' {
                    found = true;
                    break;
                }
                cursor += 1;
            }
            if !found {
                complete = false;
                diagnostics.push(Diagnostic::new(
                    "servlet_jsp.el.unclosed_comment",
                    "Unclosed comment in EL expression",
                    "warning",
                    file_path,
                    start_line,
                    start_line,
                ));
                break;
            }
            index = cursor + 2;
            continue;
        }
        let start = index;
        let kind;
        if ch == '\'' || ch == '"' {
            let quote = ch;
            index += 1;
            let mut escaped = false;
            let mut closed = false;
            while index < chars.len() {
                let current = chars[index];
                index += 1;
                if escaped {
                    escaped = false;
                } else if current == '\\' {
                    escaped = true;
                } else if current == quote {
                    closed = true;
                    break;
                }
            }
            if !closed {
                complete = false;
                diagnostics.push(Diagnostic::new(
                    "servlet_jsp.el.unclosed_string",
                    "Unclosed string literal in EL expression",
                    "warning",
                    file_path,
                    start_line,
                    start_line,
                ));
            }
            kind = "string";
        } else if ch.is_alphabetic() || ch == '_' || ch == '$' {
            index += 1;
            while index < chars.len() && (chars[index].is_alphanumeric() || chars[index] == '_' || chars[index] == '$') {
                index += 1;
            }
            kind = "identifier";
        } else if ch.is_ascii_digit() {
            index += 1;
            while index < chars.len()
                && (chars[index].is_alphanumeric() || chars[index] == '.' || chars[index] == '_')
            {
                index += 1;
            }
            kind = "number";
        } else {
            let pair: String = chars[index..(index + 2).min(chars.len())].iter().collect();
            if ["==", "!=", "<=", ">=", "&&", "||", "->", "+=", "-=", "*=", "/=", "?."].contains(&pair.as_str()) {
                index += 2;
            } else {
                index += 1;
            }
            kind = if "()[]{}.,:?".contains(ch) { "punctuation" } else { "operator" };
        }
        let value: String = chars[start..index].iter().collect();
        tokens.push(ElToken {
            kind: kind.to_string(),
            value: value.clone(),
            start_offset: start as i64 + body_offset,
            end_offset: index as i64 + body_offset,
        });
        if value == "(" || value == "[" || value == "{" {
            nesting.push(value.chars().next().unwrap());
            if nesting.len() > max_nesting {
                diagnostics.push(Diagnostic::new(
                    "servlet_jsp.el.nesting_budget",
                    &format!("EL nesting budget {max_nesting} reached"),
                    "warning",
                    file_path,
                    start_line,
                    start_line,
                ));
                truncated = true;
                break;
            }
        } else if value == ")" || value == "]" || value == "}" {
            let expected = match value.as_str() {
                ")" => '(',
                "]" => '[',
                _ => '{',
            };
            if nesting.is_empty() || *nesting.last().unwrap() != expected {
                complete = false;
                diagnostics.push(Diagnostic::new(
                    "servlet_jsp.el.unexpected_closer",
                    &format!("Unexpected closing token {value:?} in EL expression"),
                    "warning",
                    file_path,
                    start_line,
                    start_line,
                ));
            } else {
                nesting.pop();
            }
        }
        if tokens.len() >= max_tokens {
            if index < chars.len() {
                diagnostics.push(Diagnostic::new(
                    "servlet_jsp.el.token_budget",
                    &format!("EL token budget {max_tokens} reached"),
                    "warning",
                    file_path,
                    start_line,
                    start_line,
                ));
                truncated = true;
            }
            break;
        }
    }
    if !nesting.is_empty() && !truncated {
        complete = false;
        diagnostics.push(Diagnostic::new(
            "servlet_jsp.el.unbalanced_nesting",
            "Unbalanced grouping token in EL expression",
            "warning",
            file_path,
            start_line,
            start_line,
        ));
    }
    (tokens, diagnostics, truncated, complete)
}

/// Slice chuỗi theo CHAR index (khớp slicing code-point của Python).
fn char_slice(text: &str, start: i64, end: i64) -> String {
    text.chars()
        .skip(start.max(0) as usize)
        .take((end - start).max(0) as usize)
        .collect()
}

fn function_references(tokens: &[ElToken], raw: &str) -> (Vec<ElFunctionReference>, BTreeSet<usize>) {
    let mut functions: Vec<ElFunctionReference> = Vec::new();
    let mut indexes: BTreeSet<usize> = BTreeSet::new();
    if tokens.len() < 4 {
        return (functions, indexes);
    }
    for index in 0..tokens.len().saturating_sub(3) {
        let prefix = &tokens[index];
        let colon = &tokens[index + 1];
        let name = &tokens[index + 2];
        let opening = &tokens[index + 3];
        if prefix.kind == "identifier" && colon.value == ":" && name.kind == "identifier" && opening.value == "(" {
            indexes.insert(index);
            indexes.insert(index + 2);
            functions.push(ElFunctionReference {
                prefix: prefix.value.clone(),
                name: name.value.clone(),
                raw: char_slice(raw, prefix.start_offset, name.end_offset),
                start_offset: prefix.start_offset,
                end_offset: name.end_offset,
            });
        }
    }
    (functions, indexes)
}

fn references(tokens: &[ElToken], raw: &str, function_token_indexes: &BTreeSet<usize>) -> Vec<ElReference> {
    let mut references: Vec<ElReference> = Vec::new();
    for index in 0..tokens.len() {
        let token = &tokens[index];
        if token.kind != "identifier" || KEYWORDS.contains(&token.value.as_str()) || function_token_indexes.contains(&index)
        {
            continue;
        }
        let previous = if index > 0 { tokens[index - 1].value.clone() } else { String::new() };
        if previous == "." || previous == "?." || previous == ":" {
            continue;
        }
        let mut path: Vec<String> = Vec::new();
        let mut end_offset = token.end_offset;
        let mut cursor = index + 1;
        let mut dynamic = false;
        while cursor < tokens.len() {
            let marker = &tokens[cursor];
            if (marker.value == "." || marker.value == "?.")
                && cursor + 1 < tokens.len()
                && tokens[cursor + 1].kind == "identifier"
            {
                let part = &tokens[cursor + 1];
                path.push(part.value.clone());
                end_offset = part.end_offset;
                cursor += 2;
                continue;
            }
            if marker.value == "[" {
                let close = matching_bracket(tokens, cursor);
                let Some(close) = close else {
                    dynamic = true;
                    end_offset = marker.end_offset;
                    break;
                };
                let inner = &tokens[cursor + 1..close];
                if inner.len() == 1 && (inner[0].kind == "string" || inner[0].kind == "number") {
                    let index_value = literal_index(&inner[0]);
                    path.push(format!("[{index_value}]"));
                } else {
                    let raw_inner = char_slice(raw, marker.end_offset, tokens[close].start_offset)
                        .trim()
                        .to_string();
                    path.push(format!("[{raw_inner}]"));
                    dynamic = true;
                }
                end_offset = tokens[close].end_offset;
                cursor = close + 1;
                continue;
            }
            break;
        }
        let implicit_scope = implicit_scope_of(&token.value).to_string();
        let resolution_status = if !implicit_scope.is_empty() && !path.is_empty() && !dynamic {
            "resolved"
        } else if dynamic {
            "dynamic"
        } else {
            "unresolved"
        };
        references.push(ElReference {
            root: token.value.clone(),
            path,
            raw: char_slice(raw, token.start_offset, end_offset),
            start_offset: token.start_offset,
            end_offset,
            resolution_status: resolution_status.to_string(),
            implicit_scope,
        });
    }
    references
}

fn matching_bracket(tokens: &[ElToken], opening: usize) -> Option<usize> {
    let mut depth = 0i32;
    for (index, token) in tokens.iter().enumerate().skip(opening) {
        if token.value == "[" {
            depth += 1;
        } else if token.value == "]" {
            depth -= 1;
            if depth == 0 {
                return Some(index);
            }
        }
    }
    None
}

fn literal_index(token: &ElToken) -> String {
    if token.kind == "string" && token.value.chars().count() >= 2 {
        return token.value[1..token.value.len() - 1].to_string();
    }
    token.value.clone()
}

fn state_reads(references: &[ElReference]) -> Vec<ElStateRead> {
    let mut reads: Vec<ElStateRead> = Vec::new();
    for reference in references {
        if reference.implicit_scope.is_empty() {
            continue;
        }
        let mut name = String::new();
        let mut dynamic = true;
        if let Some(first) = reference.path.first() {
            if first.starts_with('[') && first.ends_with(']') {
                name = first[1..first.len() - 1].to_string();
            } else {
                name = first.clone();
            }
            dynamic = name.is_empty() || name.contains(['$', '{', '}', '[', ']', '(', ')']);
        }
        reads.push(ElStateRead {
            implicit_object: reference.root.clone(),
            scope: reference.implicit_scope.clone(),
            name,
            raw: reference.raw.clone(),
            dynamic,
            start_offset: reference.start_offset,
            end_offset: reference.end_offset,
        });
    }
    reads
}
